drivers/platform/stm32/stm32_dma.c

```diff
@@ -0,0 +1,284 @@
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Configure a DMA channel for a transfer.
+ * @param channel - The DMA channel descriptor.
+ * @param xfer - Descriptor for the transfer.
+ * @return 0 in case of success, negative error codes otherwise
+ */
+int stm32_dma_config_xfer(struct no_os_dma_ch *channel,
+			  struct no_os_dma_xfer_desc *xfer)
+{
+	struct stm32_dma_channel* sdma_ch;
+	int ret;
+
+	if (!channel || !xfer || !channel->extra)
+		return -EINVAL;
+
+	sdma_ch = channel->extra;
+
+	sdma_ch->hdma->Init.Channel = sdma_ch->ch_num;
+	sdma_ch->hdma->Init.MemInc = sdma_ch->mem_increment ? DMA_MINC_ENABLE :
+				     DMA_MINC_DISABLE;
+	sdma_ch->hdma->Init.PeriphInc = sdma_ch->per_increment ? DMA_PINC_ENABLE :
+					DMA_PINC_DISABLE;
+	sdma_ch->hdma->Init.MemDataAlignment = sdma_ch->mem_data_alignment;
+	sdma_ch->hdma->Init.PeriphDataAlignment = sdma_ch->per_data_alignment;
+	sdma_ch->hdma->Init.Mode = sdma_ch->dma_mode;
+
+	switch (sdma_ch->mem_data_alignment) {
+	case DATA_ALIGN_BYTE:
+		sdma_ch->hdma->Init.MemDataAlignment = DMA_MDATAALIGN_BYTE;
+		break;
+	case DATA_ALIGN_HALF_WORD:
+		sdma_ch->hdma->Init.MemDataAlignment = DMA_MDATAALIGN_HALFWORD;
+		break;
+	case DATA_ALIGN_WORD:
+		sdma_ch->hdma->Init.MemDataAlignment = DMA_MDATAALIGN_WORD;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (sdma_ch->per_data_alignment) {
+	case DATA_ALIGN_BYTE:
+		sdma_ch->hdma->Init.PeriphDataAlignment = DMA_PDATAALIGN_BYTE;
+		break;
+	case DATA_ALIGN_HALF_WORD:
+		sdma_ch->hdma->Init.PeriphDataAlignment = DMA_PDATAALIGN_HALFWORD;
+		break;
+	case DATA_ALIGN_WORD:
+		sdma_ch->hdma->Init.PeriphDataAlignment = DMA_PDATAALIGN_WORD;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (sdma_ch->dma_mode) {
+	case DMA_NORMAL_MODE:
+		sdma_ch->hdma->Init.Mode = DMA_NORMAL;
+		break;
+	case DMA_CIRCULAR_MODE:
+		sdma_ch->hdma->Init.Mode = DMA_CIRCULAR;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (xfer->xfer_type) {
+	case MEM_TO_MEM:
+		sdma_ch->hdma->Init.Direction = DMA_MEMORY_TO_MEMORY;
+		break;
+	case MEM_TO_DEV:
+		sdma_ch->hdma->Init.Direction = DMA_MEMORY_TO_PERIPH;
+		break;
+	case DEV_TO_MEM:
+		sdma_ch->hdma->Init.Direction = DMA_PERIPH_TO_MEMORY;
+		break;
+	default:
+		break;
+	}
+
+	sdma_ch->src = xfer->src;
+	sdma_ch->dst = xfer->dst;
+	sdma_ch->length = xfer->length;
+
+	ret = HAL_DMA_Init(sdma_ch->hdma);
+	if (ret != HAL_OK)
+		return ret;
+
+	return 0;
+}
+
+/**
+ * @brief Initialize a DMA controller.
+ * @param desc - Descriptor to be initialized.
+ * @param param - Initialization parameter for the decriptor.
+ * @return 0 in case of success
+ * 	   -ENOMEM if there is not enough free memory
+ */
+int stm32_dma_init(struct no_os_dma_desc** desc,
+		   const struct no_os_dma_init_param* param)
+{
+
+	struct no_os_dma_desc* descriptor;
+	int ret;
+	uint8_t i;
+
+	if (!desc)
```

Comment:
The current DMA driver for maxim uses shared references for `no_os_dma_desc` descriptors (meaning that if you call the dma_init function with the same `id` field you'll get the same `no_os_dma_desc *` pointer). Is there a reason for not implementing the same here?

Reply:
I am not sure I got this right..- what is the shared references you are referring to here?